orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/ManifestEvaluator.java

```diff
@@ -16,172 +16,245 @@
     }
 
-    List<Artifact> requiredArtifacts = new ArrayList<>();
-    for (String id : Optional.ofNullable(context.getRequiredArtifactIds()).orElse(emptyList())) {
-      Artifact requiredArtifact = artifactUtils.getBoundArtifactForId(stage, id);
-      if (requiredArtifact == null) {
-        throw new IllegalStateException(
-            "No artifact with id '" + id + "' could be found in the pipeline context.");
-      }
+    return getSpelEvaluatedManifests(unevaluatedManifests, stage);
+  }
 
-      requiredArtifacts.add(requiredArtifact);
+  private ImmutableList<Map<Object, Object>> coerceManifestToList(Object manifest) {
+    if (manifest instanceof List) {
+      return ImmutableList.copyOf(
+          objectMapper.convertValue(manifest, new TypeReference<List<Map<Object, Object>>>() {}));
     }
+    Map<Object, Object> singleManifest =
+        (Map<Object, Object>) objectMapper.convertValue(manifest, Map.class);
+    return ImmutableList.of(singleManifest);
+  }
 
-    // resolve SpEL expressions in artifacts defined inline in the stage
-    for (ManifestContext.BindArtifact artifact :
-        Optional.ofNullable(context.getRequiredArtifacts()).orElse(emptyList())) {
-      Artifact requiredArtifact =
-          artifactUtils.getBoundArtifactForStage(
-              stage, artifact.getExpectedArtifactId(), artifact.getArtifact());
-
-      if (requiredArtifact == null) {
-        throw new IllegalStateException(
-            "No artifact with id '"
-                + artifact.getExpectedArtifactId()
-                + "' could be found in the pipeline context.");
+  private Supplier<Iterable<Object>> fetchAndParseManifestYaml(Artifact manifestArtifact) {
+    return () -> {
+      Response manifestText = oortService.fetchArtifact(manifestArtifact);
+      try (InputStream body = manifestText.getBody().in()) {
+        return yamlParser.get().loadAll(body);
+      } catch (Exception e) {
+        throw new IllegalStateException(e);
       }
+    };
+  }
+
+  private Artifact getManifestArtifact(Stage stage, ManifestContext context) {
+    Artifact manifestArtifact =
+        Optional.ofNullable(
+                artifactUtils.getBoundArtifactForStage(
+                    stage, context.getManifestArtifactId(), context.getManifestArtifact()))
+            .orElseThrow(() -> new IllegalArgumentException("No manifest artifact was specified."));
 
-      requiredArtifacts.add(requiredArtifact);
+    // Once the legacy artifacts feature is removed, all trigger expected artifacts will be
+    // required to define an account up front.
+    if (context.getManifestArtifactAccount() != null) {
+      manifestArtifact.setArtifactAccount(context.getManifestArtifactAccount());
     }
 
-    log.info("Artifacts {} are bound to the manifest", requiredArtifacts);
+    if (manifestArtifact.getArtifactAccount() == null) {
+      throw new IllegalArgumentException("No manifest artifact account was specified.");
+    }
```

Comment:
equivalent code:
`checkArgument(manifestArtifact.getArtifactAccount() != null, "No manifest artifact account was specified.");`

Internally we use the [`Preconditions`](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/base/Preconditions.html) methods in practically every file, but maybe it would be confusing inside Spinnaker so isn't worth using, I don't know...

Reply:
Love it, adding to the plumpstuff java bible.